### DISCUSSION TITLE
Vertical and horizontal justify layout

### DISCUSSION THREAD

**Discussion by zohnannor** [emilk/egui#1409] [Q&A] [answered]
How can I place three (or possibly more) widgets (in my case, vertical-centered label + [`ScrollArea`] beneath it (which contains a multiline [`TextEdit`])) so that they'd fill space both horizontally (i.e. divide left-to-right space equally) and vertically, taking all the height available to them?

Illustration:

![image](https://user-images.githubusercontent.com/35764628/159691548-191325b0-eb5f-4d30-9c66-27a2e09bf2b9.png)

It should, of course, be responsive with resizing and stuff (that is, [`ScrollArea`]s should show scrollbar if text in `TextEdit` cannot fit vertically).

What I've tried:

- [`Ui::horizontal`]
- [`Ui::horizontal_top`]
- [`Ui::with_layout`] + different [`Layout`]s
- [`Rect`] + [`Ui::put`]

From what I've seen in the [demo] (namely, "Layout Test" window), I think egui isn't capable of justifying widgets horizontally and vertically at the same time.  
However, my understanding of positioning of widgets on layout in egui may be incorrect and I was using wrong combinations of methods listed above.

I can add more code samples and screenshots if that would help.

[`ScrollArea`]: https://docs.rs/egui/latest/egui/containers/struct.ScrollArea.html#method.vertical
[`TextEdit`]: https://docs.rs/egui/latest/egui/widgets/struct.TextEdit.html#method.code_editor
[`Ui::horizontal`]: https://docs.rs/egui/latest/egui/struct.Ui.html#method.horizontal
[`Ui::horizontal_top`]: https://docs.rs/egui/latest/egui/struct.Ui.html#method.horizontal_top
[`Ui::with_layout`]: https://docs.rs/egui/latest/egui/struct.Ui.html#method.with_layout
[`Ui::put`]: https://docs.rs/egui/latest/egui/struct.Ui.html#method.put
[`Layout`]: https://docs.rs/egui/latest/egui/struct.Layout.html#constructors
[`Rect`]: https://docs.rs/egui/latest/egui/struct.Rect.html#associatedconstant.EVERYTHING
[demo]: https://egui.rs/

**Comment by emilk** [comment, marked as answer]
As mentioned in the docs, layouting in immediate mode is difficult. If you know that you are going to divide the space in three equal parts, I suggest doing so manually. That is, take the available space (`ui.max_rect()` if this is the first thing in the `Ui`), create three `Rect`:s that divide the space in three, and call `ui.allocate_ui_at_rect` on each such `Rect`.

**Reply by zohnannor** [reply]
Thanks, got it working by manually computing the width for widgets and substracting `ui.spacing().item_spacing.x` from it, then `.translate`ing rects by that amount and using `ui.allocate_ui_at_rect` to placethe widgets.

```rust
let item_spacing = ui.spacing().item_spacing.x;
let mut max_rect = ui.max_rect();                         // get max rect size
max_rect.set_width(max_rect.width() - 2. * item_spacing); // adjust it so it does not include spacing
let width = max_rect.width() / 3.;                        // get width for the widgets
let offset = vec2(width + item_spacing, 0.);              // compute the offset for subsequent rects
max_rect.set_width(width);                                // and set the width of rect
/* placing my widget here */
max_rect = max_rect.translate(offset);
/* placing my widget here */
max_rect = max_rect.translate(offset);
/* placing my widget here */

```

Phew!

**Reply by emilk** [reply]
Good job! `ui.columns(…)` does something similar btw.

**Reply by zohnannor** [reply]
I feel so ashamed... How could I overlook `columns`?... You saved me! I was very not proud of the code above 😅  
Columns is EXACTLY what I was looking for, thanks!

**Reply by XertroV** [reply]
Found this via google and thought I'd contribute my solution -- I wanted a row of multiple labeled values that were spread out nicely (for a dashboard type thing)

```rust
// horizontal justified + centered global data
ui.horizontal(|ui| {
    ui.columns(5, |cols| {
        cols[0].vertical_centered_justified(|ui| ui.label(format!("Label0 {}", data[0])));
        cols[1].vertical_centered_justified(|ui| ui.label(format!("Label1 {}", data[1])));
        cols[2].vertical_centered_justified(|ui| ui.label(format!("Label2 {}", data[2])));
        cols[3].vertical_centered_justified(|ui| ui.label(format!("Label3 {}", data[3])));
        cols[4].vertical_centered_justified(|ui| ui.label(format!("Label4 {}", data[4])));
    });
});
```

**Comment by modulomedito** [comment]
New to egui and rust. Inspired by @zohnannor, I encapsulated the codes into a implementation which easier to place widgets into grids. It works well for me.

The final output is look like this:
<img width="591" alt="image" src="https://github.com/emilk/egui/assets/38747395/5d36137d-99a6-47f6-a7e7-bec95a83b179">

**Definition** (placed in a file named "allocate.rs")
```rust
use eframe::egui;

pub struct Grid {
    cell: egui::Rect,
    row_num: usize,
    col_num: usize,
}

impl Grid {
    pub fn new(
        ui: &mut egui::Ui,
        split_row_num: usize,
        split_col_num: usize,
    ) -> Self {
        // Get item spacing
        let item_spacing = ui.spacing().item_spacing.x;

        // Get max rect size
        let mut max_rect = ui.max_rect();

        // Adjust it so it does not include spacing
        max_rect.set_width(max_rect.width() - 2.*item_spacing);

        // Get width for the widgets
        let width = max_rect.width()/(split_col_num as f32);

        // Get height for the widgets
        let height = max_rect.height()/(split_row_num as f32);

        // Compute the offset for subsequent rects
        let offset = egui::vec2(width + item_spacing, 0.);

        // And set the width of rect
        max_rect.set_width(width);

        // And set the height of rect
        max_rect.set_height(height);

        // Return
        Self {
            cell: max_rect,
            row_num: split_row_num,
            col_num: split_col_num,
        }
    }

    pub fn place<R>(
        &self, 
        ui: &mut egui::Ui,
        row: usize,
        col: usize,
        add_ctx: impl FnOnce(&mut egui::Ui) -> R,
    ) -> egui::InnerResponse<R> {
        assert!(row >= 1);
        assert!(col >= 1);
        let height = self.cell.height();
        let width = self.cell.width();
        let colf32: f32 = (col - 1) as f32;
        let rowf32: f32 = (row - 1) as f32;
        let hold_rect = self.cell.translate(
            egui::vec2(colf32*width, rowf32*height));
        // Return
        ui.allocate_ui_at_rect(hold_rect, add_ctx)
    }
}
```

**Usage**
```rust
pub fn update(&mut self, ui: &mut egui::Ui) {
    // Create a 10 rows and 3 columns grid in current UI
    let griding = allocate::Grid::new(ui, 10, 3);
    
    // Place a widget on [4,1] cell (index starts from 1)
    griding.place(ui, 4, 1, |ui| {
        ui.centered_and_justified(|ui| {
            ui.label("row4, col1");
        });
    });
    
    // Place a widget on [3,2] cell
    griding.place(ui, 3, 2, |ui| {
        ui.centered_and_justified(|ui| {
            ui.label("row3, col2");
        });
    });
    
    // Place a widget on [5,3] cell (index starts from 1)
    griding.place(ui, 5, 3, |ui| {
        ui.centered_and_justified(|ui| {
            ui.label("row5, col3");
        });
    });
}
```
